[libclc] Declare workitem built-ins in clc, move ptx-nvidiacl workitem built-ins into clc

wenju-he opened the pull request:
Changes in this PR:
* Declare most of workitem functions in clc and opencl folders.
* Call clc workitem function in corresponding OpenCL workitem function.
* Move ptx-nvidiacl workitem built-in implementations into clc.
* Move a few amdgcn workitem built-in implementations into clc.
* Include only needed headers in OpenCL workitem functions.
* Implement get_local_linear_id, get_max_sub_group_size, get_num_sub_groups,
  get_sub_group_id, get_sub_group_local_id, get_sub_group_size for ptx-nvidiacl.

llvm-diff shows this PR adds a few new symbols to nvptx64--nvidiacl.bc.
llvm-diff shows no change to amdgcn--amdhsa.bc, nvptx--.bc and nvptx64--.bc.

wenju-he committed:
[libclc] Declare workitem built-ins in clc, move ptx-nvidiacl workitem built-ins into clc

Changes in this PR:
* Declare all workitem functions in clc and opencl folders.
* Call clc workitem function in corresponding OpenCL workitem function.
* Move ptx-nvidiacl workitem built-in implementations into clc.
* clc_get_global_offset, clc_get_enqueued_num_sub_groups,
  clc_get_enqueued_local_size and clc_get_work_dim are not implemented
  because it is not possible to provide generic implementation.
* Include only needed headers in OpenCL workitem functions.
* Add a FIXME to OpenCL get_global_id.cl.

Note opencl/lib/ptx-nvidiacl/workitem/get_global_id.cl isn't changed
because it has different implementation from generic.
llvm-diff shows this PR only adds a dozen new symbols to
amdgcn--amdhsa.bc and nvptx64--nvidiacl.bc.

wenju-he committed:
add __ prefix to clc functions

wenju-he commented on libclc/clc/include/clc/workitem/clc_get_enqueued_local_size.h:
```diff
@@ -0,0 +1,16 @@
+
+#include <clc/internal/clc.h>
+
+_CLC_OVERLOAD _CLC_DECL size_t clc_get_enqueued_local_size(uint dim);
```
> Should this (and the rest) be prefixed with `__`? Like `__clc_get_enqueued_local_size`?

done in 3bab4ee8c7ff956fafac13975ca00a715b8ba460, thanks

frasercrmck commented:
Is an issue here perhaps that downstream users may have implemented (e.g.) `get_global_offset` in their toolchains, but after this change they'll suddenly see a definition of this function that they're not expecting?

I suppose they would have to implement CLC functions, like `__clc_get_global_offset` to work with this approach. Or they could have their libclc target not call into `__clc_get_global_offset`, or just not provide a definition of `get_global_offset` as before. Either way this has the potential to be a breaking change.

I'm not sure about leaving dangling CLC declarations in OpenCL libraries. It feels like an implementation detail we're exposing to users.

wenju-he commented:
> Is an issue here perhaps that downstream users may have implemented (e.g.) `get_global_offset` in their toolchains, but after this change they'll suddenly see a definition of this function that they're not expecting?
> 
> I suppose they would have to implement CLC functions, like `__clc_get_global_offset` to work with this approach. Or they could have their libclc target not call into `__clc_get_global_offset`, or just not provide a definition of `get_global_offset` as before. Either way this has the potential to be a breaking change.
> 
> I'm not sure about leaving dangling CLC declarations in OpenCL libraries. It feels like an implementation detail we're exposing to users.

thanks @frasercrmck. As to get_work_dim and above mentioned get_global_offset, I see there are amdgpu implementations https://github.com/llvm/llvm-project/blob/main/libclc/opencl/lib/amdgcn/workitem/get_global_offset.cl and https://github.com/llvm/llvm-project/blob/main/libclc/opencl/lib/r600/workitem/get_global_offset.cl. Moving them into clc should resolve the concern, but I don't know where is destination directory in clc. I don't know how to map `opencl/lib/amdgcn` and `opencl/lib/r600` to `clc/lib/amdgcn` and `clc/lib/amdgpu`. Do you have suggestions about the mapping?

As to clc_get_enqueued_num_sub_groups and clc_get_enqueued_local_size that there is definition in this repo, I'll delete them from this PR. The motivation of adding them in the first commit is to avoid adding them in the downstream only since they are shared code.

arsenm commented:
> Moving them into clc should resolve the concern, but I don't know where is destination directory in clc. I don't know how to map `opencl/lib/amdgcn` and `opencl/lib/r600` to `clc/lib/amdgcn` and `clc/lib/amdgpu`. Do you have suggestions about the mapping?

r600 and amdgcn should be treated as separate targets. we should stop trying to have common code under an "amdgpu" name

wenju-he committed:
move a few amdgcn functions into clc to add definition of __clc_get_work_dim/__clc_get_global_offset

wenju-he commented:
> I'm not sure about leaving dangling CLC declarations in OpenCL libraries. It feels like an implementation detail we're exposing to users.

this issue has been fixed. There is no unresolved __clc symbol in final OpenCL library.

I moved a few amdgcn workitem builtins from opencl/lib/amdgcn/workitem to clc/lib/amdgcn/workitem/ so that a few newly added header file has use. llvm-diff shows no change to amdgcn--amdhsa.bc.

I moved nvptx64--nvidiacl workitem builtin to clc and added a few new workitem built-ins for nvptx64--nvidiacl so that newly added header file has use.

frasercrmck commented:
I am seeing unresolved CLC functions in `nvptx--.bc` and `nvptx64--.bc`. I think it's because we're now building things like `get_num_groups` for `nvptx`/`nvptx64`, whereas previously they were not being built for those targets. We're also not building `__clc_get_num_groups` because we're (correctly) not including `ptx-nvidiacl` sources.

wenju-he committed:
move get_group_id.cl, get_local_id.cl and get_num_groups.cl to ptx-nvidiacl, revert change to opencl/lib/generic/workitem/get_global_size.cl

wenju-he commented:
> I am seeing unresolved CLC functions in `nvptx--.bc` and `nvptx64--.bc`. I think it's because we're now building things like `get_num_groups` for `nvptx`/`nvptx64`, whereas previously they were not being built for those targets. We're also not building `__clc_get_num_groups` because we're (correctly) not including `ptx-nvidiacl` sources.

thanks @frasercrmck for the careful review. `nvptx--.bc` and `nvptx64--.bc` issue is fixed in b1397a4c0ba63a2fd2392aa064bfa351ce014a7d

Please let me know if I should revert following style change and follow existing tyle.
1.  add #ifndef __CLC_OPENCL_WORKITEM_* guard in opencl header file and include needed #include <clc/internal/clc.h>. E.g. libclc/opencl/include/clc/opencl/workitem/get_global_id.h. 
2. include only needed headers in OpenCL lib files, e.g. in libclc/opencl/lib/ptx-nvidiacl/workitem/get_local_linear_id.cl
```
#include <clc/opencl/workitem/get_local_linear_id.h>
#include <clc/workitem/clc_get_local_linear_id.h>

_CLC_OVERLOAD _CLC_DEF size_t get_local_linear_id() {
  return __clc_get_local_linear_id();
}
```

frasercrmck commented:
> > I am seeing unresolved CLC functions in `nvptx--.bc` and `nvptx64--.bc`. I think it's because we're now building things like `get_num_groups` for `nvptx`/`nvptx64`, whereas previously they were not being built for those targets. We're also not building `__clc_get_num_groups` because we're (correctly) not including `ptx-nvidiacl` sources.
> 
> thanks @frasercrmck for the careful review. `nvptx--.bc` and `nvptx64--.bc` issue is fixed in [b1397a4](https://github.com/llvm/llvm-project/commit/b1397a4c0ba63a2fd2392aa064bfa351ce014a7d)
> 
> Please let me know if I should revert following style change and follow existing tyle.
> 
>     1. add #ifndef __CLC_OPENCL_WORKITEM_* guard in opencl header file and include needed #include <clc/internal/clc.h>. E.g. libclc/opencl/include/clc/opencl/workitem/get_global_id.h.
> 
>     2. include only needed headers in OpenCL lib files, e.g. in libclc/opencl/lib/ptx-nvidiacl/workitem/get_local_linear_id.cl
> 
> 
> ```
> #include <clc/opencl/workitem/get_local_linear_id.h>
> #include <clc/workitem/clc_get_local_linear_id.h>
> 
> _CLC_OVERLOAD _CLC_DEF size_t get_local_linear_id() {
>   return __clc_get_local_linear_id();
> }
> ```

I think that's a good approach, thanks. We generally include far too much stuff in the OpenCL layer.

One comment is that I had initially intended for `clc/internal` to only be used by the CLC builtins. I realise it's already being used by some `fma.cl`s but we might want to nip it in the bud. Perhaps we need a stripped down OpenCL header, like `clc/opencl/opencl-base.h` which includes just the types and macro defs?

frasercrmck commented:
> Perhaps we need a stripped down OpenCL header, like `clc/opencl/opencl-base.h` which includes just the types and macro defs?

See #146840 for an example of this approach.

wenju-he committed:
Merge branch 'main' into clc-workitem

wenju-he committed:
include clc/opencl/opencl-base.h in opencl headers

arsenm commented on libclc/clc/lib/amdgcn/workitem/clc_get_global_size.cl:
```diff
@@ -0,0 +1,22 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include <clc/workitem/clc_get_global_size.h>
```
Why are all of these using `#include <>` instead of `#include ""`? Probably should do a bulk fix of these at some point

wenju-he replied:
> Why are all of these using `#include <>` instead of `#include ""`? Probably should do a bulk fix of these at some point

it is following existing style. I'll do the bulk fix.

arsenm commented on libclc/clc/lib/amdgcn/workitem/clc_get_global_size.cl:
```diff
@@ -0,0 +1,22 @@
+  case 2:
+    return __builtin_amdgcn_grid_size_z();
+  default:
+    return 1;
```
I thought this was supposed to be UB

wenju-he replied:
> I thought this was supposed to be UB

The code is moved as it is from https://github.com/llvm/llvm-project/blob/main/libclc/opencl/lib/amdgcn/workitem/get_global_size.cl and OpenCL spec says `For other values of dimindx, get_global_size() returns 1.`

Do you mean the implementation in clc should change to be __builtin_unreachable?

wenju-he committed:
Merge branch 'main' into clc-workitem
